power_grid_model_c/power_grid_model/include/power_grid_model/auxiliary/serialization/deserializer.hpp

```diff
@@ -739,49 +744,67 @@ class Deserializer {
         offset_ = 0;
     }
 
-    void parse_map_element(void* element_pointer, Idx map_size, MetaComponent const& component) {
+    void parse_map_element(BufferView const& buffer, Idx map_size, MetaComponent const& component) {
         while (map_size-- != 0) {
             attribute_key_ = parse_string();
-            Idx const found_idx = component.find_attribute(attribute_key_);
-            if (found_idx < 0) {
+            Idx const component_idx = component.find_attribute(attribute_key_);
+            if (component_idx < 0) {
                 attribute_key_ = {};
                 // allow unknown key for additional user info
                 parse_skip();
                 continue;
             }
-            parse_attribute(element_pointer, component.attributes[found_idx]);
+            parse_attribute(buffer, component, component.attributes[component_idx]);
         }
         attribute_key_ = "";
     }
 
-    void parse_array_element(void* element_pointer, Idx array_size, std::span<MetaAttribute const* const> attributes) {
+    void parse_array_element(BufferView const& buffer, Idx array_size, MetaComponent const& component,
+                             std::span<MetaAttribute const* const> attributes) {
         if (array_size != static_cast<Idx>(attributes.size())) {
             throw SerializationError{
                 "An element of a list should have same length as the list of predefined attributes!\n"};
         }
         for (attribute_number_ = 0; attribute_number_ != array_size; ++attribute_number_) {
-            parse_attribute(element_pointer, *attributes[attribute_number_]);
+            parse_attribute(buffer, component, *attributes[attribute_number_]);
         }
         attribute_number_ = -1;
     }
 
-    void parse_attribute(void* element_pointer, MetaAttribute const& attribute) {
+    void parse_attribute(BufferView const& buffer, MetaComponent const& component, MetaAttribute const& attribute) {
         // call relevant parser
-        ctype_func_selector(attribute.ctype, [element_pointer, &attribute, this]<class T> {
-            ValueVisitor<T> visitor{{}, attribute.get_attribute<T>(element_pointer)};
+        assert(buffer.buffer != nullptr);
+
+        ctype_func_selector(attribute.ctype, [&]<class T> {
+            T skip_element{};
+            ValueVisitor<T> visitor{
+                {}, [&]() -> T& {
+                    if (buffer.buffer->data != nullptr) {
+                        return attribute.get_attribute<T>(component.advance_ptr(buffer.buffer->data, buffer.idx));
+                    }
+                    if (auto it =
+                            std::ranges::find_if(buffer.buffer->attributes,
+                                                 [attribute_ptr = &attribute](AttributeBuffer<void> const& attr_buf) {
+                                                     return attr_buf.meta_attribute == attribute_ptr;
+                                                 });
```

Comment:
this is a potentially expensive search in the hot loop. Unfortunately, I have tried many different things but haven't been able to come up with a satisfactory solution except for making 2 completely separate control flows (one for row-based, one for columnar).

@TonyXiang8787 what do you think?

Reply:
# Let me start with some observations:

1. For a given element, if the element is a `dict` with verbose attribute names, the name look-up per attribute per element is not avoidable, regardless of if we have columnar or row-based format. We just need to make sure the look-up does not happen twice.
2. For a given element, if the element is a `list` with pre-defined attribute names in the header, your current implementation is very inefficient for columnar format as we still do the name look-up per attribute per element, which is not needed for row-based format.

# Now my proposal:

## Columnar buffer order

At the beginning of the `parse` function, we need to define another function called `map_columnar_attributes` or something, it should per component check if the provided buffer is columnar. If yes, it maps in one go the sequence number of the provided columnar buffers regards to the sequence of the pre-defined attributes, with `-1` meaning the attribute buffer is not provided. 

For example, for the following attributes list

```json
{
  "version": "1.0",
  "type": "sym_output",
  "is_batch": false,
  "attributes": {
       "node": ["id", "u_pu", "u_angle"]
   },
  "data": {}
}
```

If the user provide columnar buffers in the order `["u_angle", "u", "u_pu"]`, the `map_columnar_attributes` should return `{"node": [-1, 2, 1]}`.

## Parse map element

In the `parse_map_element`, if the buffer is columnar, you look-up the attribute names directly from the attribute list of provided columnar buffer. If the buffer is row-based, you look-up the attribute names from the metadata as currently.

## Parse array element

In the `parse_array_element`, if the bufer is columnar, you refer the sequence from `map_columnar_attributes` to retrieve directly the relevant buffer position without to do name look-up again.

## Parse attribute

You need two versions of `parse_attribute`, for row-based and columnar.

Reply:
i think i've processed all your comments on the deserializer. if you have time, can you do another check?

Reply:
The new implementation looks good. I hereby resolve all de-serializer comments.